File: tp1/TP01Q12 - RECURSIVO - Palíndromo em C/palindromo_recursivo.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <locale.h>
#include <stdbool.h>
#define TAM 1000 //definição de TAM para 1000 para facilitar ajustes no código

bool testar(char *frase, int dir, int esq, bool resp){ //metodo para conferir se é palindromo
	 if(esq < dir){ //condição para que esq seja menor que dir
     	if(frase[esq] != frase[dir]){ //caso os caracteres sejam diferentes resp = false e esq = tam
     		resp = false;
     		esq = dir;
     	}
     	resp = testar(frase, --dir, ++esq, resp); //chamada ao metodo decrementando dir e incrementando esq
     }	
     return resp;
}

int main(){
    char frase[TAM];
    int tam = 0, i = 1;
    bool resp = true;
    
    scanf(" %[^\n]", frase); //leitura da primeira linha do array

	if(strcmp(frase, "FIM") == 0){ //comparação do array frase com FIM que corresponde à última linha do arquivo pub.in
        i = 0;
    }

    while(i != 0){	//loop para leitura das linhas do arquivo

 		tam = strlen(frase)-1; //tamanho do array
        
        resp = testar(frase, tam, 0, true);

        if(resp == true){ //condição para printar se é palindromo ou não baseado nas comparações do for acima
            printf("SIM\n");
        }else{
            printf("NAO\n");
        }
       
        scanf(" %[^\n]", frase); //leitura de linha a linha do array
        
        if(strcmp(frase, "FIM") == 0){ //condição para ver se a frase lida corresponde ao final do arquivo identificado por FIM
            i = 0;
        }
        
    }
       
    return 0;
}

File: tp1/TP01Q11 - RECURSIVO - Palíndromo em Java/palindromo_recursivo.java
import java.io.*;

class palindromo_recursivo{
	public static void main (String[] args)throws Exception{
		String frase = "";
		boolean resp;
		int i = 0, tam = 0;
		
		while(i != 1){ //loop para leitura de todo arquivo
			frase = MyIO.readLine();
			if(frase.charAt(0) == 'F' && frase.charAt(1) == 'I' && frase.charAt(2) == 'M'){  //FIM presente no fim do arquivo
				i = 1; 																		
			}else{			
				tam = frase.length(); //tamanho do array frase
				tam = tam - 1;
            	resp = testar(frase, tam);
            	
            	if(resp == true){ //condição para printar na tela se é palindromo ou não
                	MyIO.println("SIM");
            	}else{
                	MyIO.println("NAO");
            	}
			}
		}	
	}
	
	public static boolean testar(String frase, int tam)throws Exception{ //metodo para chamar outro metodo com dois parametros a mais
         return testar(frase, tam, 0, true);
	}
	
	public static boolean testar(String frase, int tam, int i, boolean resp)throws Exception{ //metodo para testar se é palindromo
		if(i < tam){ //condição para que o i não seja igual ou maior que tam
			if(frase.charAt(i) != frase.charAt(tam)){ //caso esses caracteres sejam diferentes i = tam e resp é = false
				resp = false;
				i = tam;
			}
			resp = testar(frase, --tam, ++i, resp); //chamar novamente o metodo incrementando o i e decrementando o tam
		}
		return resp;
	}
}









